### PR TITLE
Use hailctl for starting Google clusters and submitting Hail jobs

### PR DESCRIPTION
**Pull request by ClintAtTheBroad**
Recently, Google changed the VM images that we'd been using to make clusters for running Hail, such that the Google- and Hail-team-supplied cluster-init scripts that we'd been using no longer worked.  This prevented running Hail jobs, since there was no feasibly-discoverable combination of images and init scripts that would make a cluster capable of running Hail.

The Hail team couldn't provide such a combination either, unfortunately.  They recommended using Hail's own wrapper around `gcloud` that takes care of the extremely fiddly parts of cluster-setup and Hail-job submission. 

This branch switches LS to using `hailctl` for Google cluster creation and Hail job submission, instead of `gcloud`.  It re-uses existing code (that invokes `gcloud`) to delete clusters and check cluster status, since `hailctl` isn't needed and doesn't easily support those use cases, respectively.  (Note that `hailctl` is itself a python wrapper around `gcloud`.)

`hailctl` submits Hail jobs that run the same version of Hail on the cluster as the Hail installation that provides `hailctl`.  This presented a problem if we wanted to maintain the ability to determine the Hail version to use via a config file change only.  The solution I came up with is to install different versions of Hail side-by-side in (mini)conda environments, and to specify which environment to activate in `loamstream.conf`.

Changes:
- Use `hailctl` for cluster creation and job submission.  See `HailCtlDataProcClient` and `HailSupport` for cluster creation and job submission commands, respectively.
- `CloudSdkDataProcClient` is now `CloudSdkDataProcWrapper`, since it no longer starts clusters, and doesn't implement the full `DataProcClient` API.
- `metadata`, `initializationActions`, `imageVersion`, and `scopes` are no longer options in `loamstream.googlecloud` in `loamstream.conf`.
- `jar` and `zip` are no longer options in `loamstream.googlecloud.hail` in `loamstream.conf`. 
- The conda environment that provides Hail is specified using `loamstream.googlecloud.hail.condaEnv`.  This field is required.
- The Google/Hail integration test uses the new setup and runs successfully.

### CODE REVIEW

File: src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+    val exitCode = runCommand(deleteClusterTokens(config))
+
+    ExitCodes.throwIfFailure(exitCode)
+  }
+
+  override def isClusterRunning: Boolean = {
+    val exitCode = runCommand(isClusterRunningTokens(config))
+
+    val result = ExitCodes.isSuccess(exitCode)
```
Comment (massung) [inline src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala]: So, this bit of code took me a bit to grok. It's asking for `isClusterRunning`, but then checking `exitCode` for success, which instantly made me think it's done running. After re-examining and looking at the surrounding functions, it becomes more clear what's going on: run an external command to check for status and then testing the exit code of the external command.

I just wonder if this could be done a little differently to make it a bit more clear at first glance? Possible ideas (not trying to indicate they're better, just throwing stuff at the wall...):

Make the `runCommand` more prominent as a `match`?

```scala
runCommand(isClusterRunningTokens(config)) match {
  case Success(_) => true
  case _ => false
}
```

Allow the return type of `runCommand` to specify whether or not it was successful?

```scala
runCommand(isClusterRunningTokens(config)).isSuccess
```

If the latter is appealing, I wonder if it would be more clear to allow the tokens to run themselves then?

```scala
isClusterRunningTokens(config).run().isSuccess
```

Anyway, just tossing out ideas.

File: src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+object CloudSdkDataProcWrapper extends Loggable {
+  def fromConfig(config: GoogleCloudConfig): Try[CloudSdkDataProcWrapper] = {
+    val gcloudBinary = config.gcloudBinary.toFile
+
+    if (gcloudBinary.exists && gcloudBinary.canExecute) {
+      Success(new CloudSdkDataProcWrapper(config))
+    } else {
+      Tries.failure(s"gcloud executable not found at ${config.gcloudBinary} or not executable")
```
Comment (massung) [inline src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala]: How is `Tries.failure` different from `Failure`? Confused a bit since `Try` is in the standard library and `Tries` is not?

File: src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    }
+  }
+
+  private[googlecloud] def deleteClusterTokens(config: GoogleCloudConfig): Seq[String] = {
+    gcloudTokens(config)("delete")(config.clusterId)
```
Comment (massung) [inline src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala]: Just wondering... these "tokens" functions look like they are "constant" for the config and only use a `GoogleCloudConfig` in order to produce the them. Why are these functions not part of `GoogleCloudConfig` and maybe even just a (lazy?) `val` member?

Just wondering if there's some rational behind that would impact later code or that I should be aware of for the future.

File: src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    gcloudTokens(config)("describe")(config.clusterId)
+  }
+  
+  private[googlecloud] def gcloudTokens(config: GoogleCloudConfig)(verb: String)(args: String*): Seq[String] = {
+    val gcloud = normalize(config.gcloudBinary)
+
+    gcloud +: "beta" +: "dataproc" +: "clusters" +: verb +: "--project" +: config.projectId +: args
+  }
+
+  private def runCommand(tokens: Seq[String]): Int = runProcess(tokens.mkString(" "), "Google Cloud SDK")
```
Comment (massung) [inline src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala]: Just noting from previous comment that now that I see tokens is `Seq[String]` and so something like `tokens.runCommand()` wouldn't be possible unless it was its own type... which might not be the worst idea for type-safety?

```scala
final case class GCloudTokens(tokens: Seq[String])
```

File: src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    debug(s"Running ${logPrefix} command: '${commandLine}'")
+    
+    import scala.sys.process._
+    
+    // STDERR messages aren't logged as 'error' because 'gcloud' writes a lot of non-error messages to STDERR
+    val processLogger = ProcessLogger(
+      line => info(s"${logPrefix}: ${line}"),
+      line => info(s"${logPrefix} (via stderr): ${line}"))
+
+    val processBuilder = BashScript.fromCommandLineString(commandLine).processBuilder(Paths.get("."))
```
Comment (massung) [inline src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala]: Just curious, does this run remotely or locally?

File: src/main/scala/loamstream/googlecloud/DataProcClient.scala [hunk 6]
```diff
@@ -4,10 +4,20 @@ package loamstream.googlecloud
  * @author clint
  * Nov 28, 2016
  */
-trait DataProcClient {
-  def startCluster(): Unit
+trait DataProcClient extends DataProcClient.CanStart with DataProcClient.CanStop with DataProcClient.CanTellIfRunning
+
+object DataProcClient {
```
Comment (massung) [inline src/main/scala/loamstream/googlecloud/DataProcClient.scala]: The traits defined in here seem like they'd be very useful outside of the `googlecloud` package and instead as part of some generic `cloud.google` package (thinking ahead to when you add AWS support). They are pretty generic.

File: src/main/scala/loamstream/googlecloud/HailCtlDataProcClient.scala [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  def fromConfigs(googleConfig: GoogleCloudConfig, hailConfig: HailConfig): Try[HailCtlDataProcClient] = {
+    for {
+      wrapper <- CloudSdkDataProcWrapper.fromConfig(googleConfig)
+    } yield {
+      new HailCtlDataProcClient(googleConfig, hailConfig, new CloudSdkDataProcWrapper(googleConfig))
+    }
+  }
+
+  private[googlecloud] def startClusterTokens(config: GoogleCloudConfig): Seq[String] = {
+    val tokens: Seq[String] = Seq(
```
Comment (massung) [inline src/main/scala/loamstream/googlecloud/HailCtlDataProcClient.scala]: zomg... This would likely be my first submit pass as well. Here's hoping in the future these tokens can all be done another way since I'm sure many of these are common among all commands (e.g. `--project` and `--zone`) and typos suck.

```scala
val projectToken = Seq("--project", projectId)
val zoneToken = Seq("--zone", zone)
val clusterToken = Seq(clusterId)
// ...

def token(args: Seq[Seq[String]], rest: String*) = args.flatten ++ rest
```

File: src/test/scala/loamstream/googlecloud/CloudSdkDataProcWrapperTest.scala [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+  test("gcloudTokens") {
+    import CloudSdkDataProcWrapper.gcloudTokens
+
+    val tokens = gcloudTokens(config)("foo")("--bar", "Baz")
+
+    val expectedTokens = {
+      Seq(examplePath.render, "beta", "dataproc", "clusters", "foo", "--project", config.projectId, "--bar", "Baz")
+    }
+    
+    assert(tokens === expectedTokens)
```
Comment (massung) [inline src/test/scala/loamstream/googlecloud/CloudSdkDataProcWrapperTest.scala]: Since command line arguments are (for the most part) rarely order dependent, I wonder if this might be better done as a `Set` and just confirming that all the tokens are present instead of being a perfect-order match? Not sure off the top what the cons would be, but there are probably a couple I'm missing.

File: src/test/scala/loamstream/googlecloud/GoogleCloudChunkRunnerTest.scala [hunk 9]
```diff
@@ -504,7 +504,7 @@ final class GoogleCloudChunkRunnerTest extends FunSuite with ProvidesEnvAndResou
       assert(client.deleteClusterInvocations() === 0)
       assert(client.isClusterRunningInvocations() === 3)
       
-      client.deleteCluster()
+      client.stopCluster()
```
Comment (massung) [inline src/test/scala/loamstream/googlecloud/GoogleCloudChunkRunnerTest.scala]: Just because this code is different, I want to make sure the terminology is good? On AWS `stop != terminate/delete`, it just means relinquishing the CPU time back to AWS, but the VM is still there in a usable state.

Is stopping different from deletion or the same for google cloud? If different (like AWS), why the change here?

**Comment by massung** [comment]
When merging with master and compiling, I see a lot of dependencies were updated (👍), but also saw this line fly by:

```
[warn] There may be incompatibilities among your library dependencies; run 'evicted' to see detailed eviction warnings.
```

Everything compiled, but wanted to make sure something bad wasn't going on.

**Comment by massung** [comment]
Running the tests I get a lot of failures. I'll look at them tomorrow, but there are so many tests streaming by (from the just the `HailSupportTest` that I can't scroll back far enough to see the errors.

It would be very helpful if you could merge w/ master, run the tests, too (locally) and see if they all pass for you or not first. If they do, then I can look at typical Windows stuff: line endings and path separators.

File: src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala (hunk 2 above)
Comment (massung) [reply src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala]: Side note: when I see `Tries` I think of string trees: https://en.wikipedia.org/wiki/Trie 😄

Comment (ClintAtTheBroad) [reply src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala]: `Tries.failure` is just a helper method that makes a `Failure[A]` wrapping a particular exception with a particular error message.  This lets you write 
```scala
Tries.failure("oof!")
```
instead of the more verbose
```scala
Failure(new Exception("oof!"))
```
The name `Tries` comes from the convention of putting "static" helper methods related to `<typename>` in a container called `<plural-form-of-typename>`, say `Foo`-related methods in `Foos`.  

I _also_ think of the data structure, but `Trys` looked funny, and something like, say, `Attempts` broke convention by being a different word entirely.

File: src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala (hunk 3 above)
Comment (ClintAtTheBroad) [reply src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala]: A few things: I liked the `*Config` classes just being dumb containers of data, since that removes the possibility for surprises regarding how they're de-serialized from typesafe-config objects.  I don't know for sure about `Ficus` but it's certainly not rare for macro- and reflection-based libs like that to get confused by "other" members that aren't mapped to anything in the serialized form.

I also wanted to keep the particulars about how command-lines are built from data (or if command lines are built at all) tightly scoped to the places where those command lines are actually needed and used.  `CloudSdkDataProcWrapper` is just one implementation of a "thing that does Google dataproc cluster-management operations" that happens to build and run commands locally.  I didn't want that detail to leak out and get baked into other APIs or couplings.

The methods that make command-line tokens could be vals (lazy or otherwise).  It was just marginally easier/more straightforward to test
```scala
member(knownInput) == expectedOutput
```
instead of 
```scala
WrappingClass(knownInput).member == expectedOutput
```
That's a small thing, and those two forms _are_ functionally equivalent.

File: src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala (hunk 4 above)
Comment (ClintAtTheBroad) [reply src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala]: I see what you mean, and I'll think it over.  My gut feeling is that the benefit isn't very big here.  This would mean (maybe) one more way of running commands.  There are already 2 or 3 depending on how you count, and I'm already trying to consolidate _those_ into one way.

File: src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala (hunk 5 above)
Comment (ClintAtTheBroad) [reply src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala]: Locally.  Or, maybe "both" depending on how you look at things.  That line runs a command locally, that in practice is a Google SDK executable that causes things to happen on remotely on Google's side.

File: src/main/scala/loamstream/googlecloud/DataProcClient.scala (hunk 6 above)
Comment (ClintAtTheBroad) [reply src/main/scala/loamstream/googlecloud/DataProcClient.scala]: Agreed.  This was one of those cases where designing for testability led to decent abstractions.  We never planned to run LS at Amazon, say, but we didn't want to rule that out.  Mostly, we needed to describe operations that could happen via shelling out to CLI tools, maybe in the future calling some Java API, or in some mock way for tests.  That led to a high-level, just-the-facts API here at least.

File: src/main/scala/loamstream/googlecloud/HailCtlDataProcClient.scala (hunk 7 above)
Comment (ClintAtTheBroad) [reply src/main/scala/loamstream/googlecloud/HailCtlDataProcClient.scala]: There are definitely ways to improve this.  In `CloudSdkDataProcWrapper`, there's `gcloudTokens` which consolidates prepending common tokens for `gcloud` commands, like
```
gcloud +: "beta" +: "dataproc" +: "clusters" +: verb +: "--project" +: config.projectId
```
`hailctl` has a different CLI than `gcloud` though, so that method can't be used for `hailctl` commands.  Some _flags_ happen to be the same between both of those tools, like `--project`, but the cost/benefit of factoring out code for just that flag didn't seem high enough.

File: src/test/scala/loamstream/googlecloud/CloudSdkDataProcWrapperTest.scala (hunk 8 above)
Comment (ClintAtTheBroad) [reply src/test/scala/loamstream/googlecloud/CloudSdkDataProcWrapperTest.scala]: I hear you, but there are some wrinkles here.  _Some_ of the tokens get their meaning from their positions, like `examplePath.render, "beta", "dataproc", "clusters", "foo"`, while for flags like `--project`, order doesn't matter.

It would be possible to take the first N tokens and expect them to all be in some order, and then treat the rest as an unordered set of pairs.  That would make for a more flexible test, which is better in general, but also be a more complex test, which is worse, all other things being equal.  In this case, I went with the simpler test.

File: src/test/scala/loamstream/googlecloud/GoogleCloudChunkRunnerTest.scala (hunk 9 above)
Comment (ClintAtTheBroad) [reply src/test/scala/loamstream/googlecloud/GoogleCloudChunkRunnerTest.scala]: This is a fair point.  I went with "stop" because that's what `hailctl` calls that operation (even though it boils down to `gcloud ... delete`), and I originally planned to use `hailctl` for everything.  `stop` is also a dual of `start`.

`delete` seemed too tied to Google's terminology, but if that's what AWS calls it, then that's a good reason to use `delete`.

>Is stopping different from deletion or the same for google cloud?

`hailctl`'s "stop" is the same as `gcloud`'s "delete", which means shut down and get rid of the cluster and all the nodes in it.

**Comment by ClintAtTheBroad** [comment]
> When merging with master and compiling, I see a lot of dependencies were updated (+1), but also saw this line fly by:

Yeah, this is embarrassing.  That's been there for years, since some Google deps bring in different transitive versions of Guava (and other stuff) transitively.  Everything has always worked, which is why fixing the warning has been low-priority, but it's a good idea to not have warnings around that everyone ignores.

File: src/test/scala/loamstream/googlecloud/GoogleCloudChunkRunnerTest.scala (hunk 9 above)
Comment (massung) [reply src/test/scala/loamstream/googlecloud/GoogleCloudChunkRunnerTest.scala]: To be clear, in AWS it's called "terminate", but it's essentially "delete", "nuke", "obliterate", and otherwise "remove from existence". ;-)

**Comment by ClintAtTheBroad** [comment]
> It would be very helpful if you could merge w/ master, run the tests, too (locally) and see if they all pass for you or not first. 

Will do.

File: src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala (hunk 1 above)
Comment (ClintAtTheBroad) [reply src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala]: I'll see what I can do here.  `ExitCodes` is an old class, that attempts to abstract over checking exit codes, since at one point we were dealing with oddball tools that didn't follow the "0 means success" convention.  That doesn't apply anymore, and I've been trying to move away from `ExitCodes` and invoking `scala.sys.process` stuff directly in favor of `loamstream.util.Processes.runSync` - which would be close to what you're proposing here, but there are a few subtly-different use cases to reconcile.

Comment (ClintAtTheBroad) [reply src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala]: I'm going to defer this to another PR.  I tried a few variations here and couldn't come up with something that didn't introduce a ton of code, or provided a good bang for the buck.  Later, I'd still like to get rid of the direct invocation of `scala.sys.process` by `CloudSdkDataProcWrapper` in favor of one way to run processes provided by `loamstream.util.Processes`.

File: src/test/scala/loamstream/googlecloud/GoogleCloudChunkRunnerTest.scala (hunk 9 above)
Comment (ClintAtTheBroad) [reply src/test/scala/loamstream/googlecloud/GoogleCloudChunkRunnerTest.scala]: I'm going to keep this as `stop` just because it's the dual of `start`.  If/when we add AWS support, I'll switch this to a name that makes sense for GCP and AWS.

**Comment by ClintAtTheBroad** [comment]
> It would be very helpful if you could merge w/ master, run the tests, too (locally) and see if they all pass for you or not first.

@massung 

I did that and confirmed that the tests pass for me. I also branched off of the head of master, merged `cg-hailctl` into that branch, and had Jenkins run the unit tests on it.  They passed too, so the problems are likely Windows-related.

File: src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala (hunk 4 above)
Comment (ClintAtTheBroad) [reply src/main/scala/loamstream/googlecloud/CloudSdkDataProcWrapper.scala]: I liked this idea, but it didn't provide a lot of bang for the buck when I tried it.  :\  I'll revisit this the next time I work on this code.

**Comment by ClintAtTheBroad** [comment]
@massung, if you have time, I'd appreciate it if you could get the tests working on Windows, then merge or not as you see fit.

**Comment by ClintAtTheBroad** [comment]
Merging since @massung is out.
